File: 349_intersection_of_two_arrays.php
<?php

class Solution {

  /**
   * @param int[] $nums1
   * @param int[] $nums2
   * @return int[]
   */
  function intersection($nums1, $nums2) {
    $num1Mapping = [];

    foreach ($nums1 as $num1) {
      $num1Mapping[$num1] = true;
    }

    $intersection = [];
    $num2Mapping = [];
    foreach ($nums2 as $num2) {
      if (isset($num1Mapping[$num2]) && !isset($num2Mapping[$num2])) {
        $intersection[] = $num2;
        $num2Mapping[$num2] = true;
      }
    }

    return $intersection;
  }
}

$solution = new Solution();

var_dump($solution->intersection([1, 2, 2, 1], [2, 2])); // [2]
var_dump($solution->intersection([4, 9, 9, 5], [9, 4, 9, 8, 4])); // [9, 4]

File: 349_intersection_of_two_arrays.java
class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
      HashMap<Integer, Boolean> nums1mapping = new HashMap<>();
      for (int i = 0; i < nums1.length; i++) {
        nums1mapping.put(nums1[i], true);
      }

      List<Integer> result = new ArrayList<>();
      HashMap<Integer, Boolean> nums2mapping = new HashMap<>();

      for (int i = 0; i < nums2.length; i++) {
        if (nums1mapping.containsKey(nums2[i]) && !nums2mapping.containsKey(nums2[i])) {
          result.add(nums2[i]);
          nums2mapping.put(nums2[i], true);
        }
      }

      int[] resultArray = new int[result.size()];
      for (int i = 0; i < result.size(); i++) {
        resultArray[i] = result.get(i);
      }

      return resultArray;
    }
}
